From d2d4d6888248a15f3b49f5ff0942b2dbdd0acdb3 Mon Sep 17 00:00:00 2001
From: Andrey Semashev <andrey.semashev@gmail.com>
Date: Mon, 2 Sep 2019 21:03:52 +0300
Subject: [PATCH] Renamed strings.h to jam_strings.h to distinguish from POSIX
 strings.h.

Some system headers on POSIX systems indirectly include strings.h in extern "C"
region. This sometimes results in Boost.Build's strings.h being included into
such region, which marks all string_* functions as extern "C" and changes their
name mangling rules accordingly, which causes linking errors. To resolve this
header conflict, this commit renames strings.h to jam_strings.h. And strings.cpp
to jam_strings.cpp for consistency.

Fixes https://github.com/boostorg/build/issues/468.
---
 tools/build/src/engine/build.bat                        | 2 +-
 tools/build/src/engine/build.sh                         | 2 +-
 tools/build/src/engine/builtins.cpp                     | 2 +-
 tools/build/src/engine/class.cpp                        | 2 +-
 tools/build/src/engine/command.h                        | 2 +-
 tools/build/src/engine/compile.cpp                      | 2 +-
 tools/build/src/engine/debugger.cpp                     | 2 +-
 tools/build/src/engine/execcmd.h                        | 2 +-
 tools/build/src/engine/execunix.cpp                     | 2 +-
 tools/build/src/engine/filent.cpp                       | 2 +-
 tools/build/src/engine/filesys.cpp                      | 2 +-
 tools/build/src/engine/fileunix.cpp                     | 2 +-
 tools/build/src/engine/filevms.cpp                      | 2 +-
 tools/build/src/engine/function.h                       | 2 +-
 tools/build/src/engine/hdrmacro.cpp                     | 2 +-
 tools/build/src/engine/jam.cpp                          | 2 +-
 tools/build/src/engine/{strings.cpp => jam_strings.cpp} | 2 +-
 tools/build/src/engine/{strings.h => jam_strings.h}     | 4 ++--
 tools/build/src/engine/modules.cpp                      | 2 +-
 tools/build/src/engine/modules/order.cpp                | 2 +-
 tools/build/src/engine/modules/regex.cpp                | 2 +-
 tools/build/src/engine/pathsys.h                        | 2 +-
 tools/build/src/engine/search.cpp                       | 2 +-
 tools/build/src/engine/timestamp.cpp                    | 2 +-
 tools/build/src/engine/variable.cpp                     | 2 +-
 tools/build/src/engine/w32_getreg.cpp                   | 2 +-
 26 files changed, 27 insertions(+), 27 deletions(-)
 rename tools/build/src/engine/{strings.cpp => jam_strings.cpp} (99%)
 rename tools/build/src/engine/{strings.h => jam_strings.h} (92%)

diff --git a/tools/build/src/engine/build.bat b/tools/build/src/engine/build.bat
index e0233afed9..cd6cefac8a 100644
--- a/tools/build/src/engine/build.bat
+++ b/tools/build/src/engine/build.bat
@@ -174,7 +174,7 @@ set B2_SOURCES=%B2_SOURCES% execcmd.cpp execnt.cpp filent.cpp filesys.cpp frames
 set B2_SOURCES=%B2_SOURCES% glob.cpp hash.cpp hcache.cpp hdrmacro.cpp headers.cpp jam.cpp
 set B2_SOURCES=%B2_SOURCES% jambase.cpp jamgram.cpp lists.cpp make.cpp make1.cpp md5.cpp mem.cpp modules.cpp
 set B2_SOURCES=%B2_SOURCES% native.cpp object.cpp option.cpp output.cpp parse.cpp pathnt.cpp
-set B2_SOURCES=%B2_SOURCES% pathsys.cpp regexp.cpp rules.cpp scan.cpp search.cpp strings.cpp
+set B2_SOURCES=%B2_SOURCES% pathsys.cpp regexp.cpp rules.cpp scan.cpp search.cpp jam_strings.cpp
 set B2_SOURCES=%B2_SOURCES% subst.cpp sysinfo.cpp timestamp.cpp variable.cpp w32_getreg.cpp
 set B2_SOURCES=%B2_SOURCES% modules/order.cpp
 set B2_SOURCES=%B2_SOURCES% modules/path.cpp
diff --git a/tools/build/src/engine/build.sh b/tools/build/src/engine/build.sh
index eef9b3da92..570b779fd7 100755
--- a/tools/build/src/engine/build.sh
+++ b/tools/build/src/engine/build.sh
@@ -413,7 +413,7 @@ B2_SOURCES="\
  rules.cpp \
  scan.cpp \
  search.cpp \
- strings.cpp \
+ jam_strings.cpp \
  subst.cpp \
  sysinfo.cpp \
  timestamp.cpp \
diff --git a/tools/build/src/engine/builtins.cpp b/tools/build/src/engine/builtins.cpp
index 01f1815a38..9a008f644d 100644
--- a/tools/build/src/engine/builtins.cpp
+++ b/tools/build/src/engine/builtins.cpp
@@ -23,7 +23,7 @@
 #include "parse.h"
 #include "pathsys.h"
 #include "rules.h"
-#include "strings.h"
+#include "jam_strings.h"
 #include "subst.h"
 #include "timestamp.h"
 #include "variable.h"
diff --git a/tools/build/src/engine/class.cpp b/tools/build/src/engine/class.cpp
index 1a06deb3e2..ec71f93ad6 100644
--- a/tools/build/src/engine/class.cpp
+++ b/tools/build/src/engine/class.cpp
@@ -13,7 +13,7 @@
 #include "lists.h"
 #include "object.h"
 #include "rules.h"
-#include "strings.h"
+#include "jam_strings.h"
 #include "variable.h"
 #include "output.h"

diff --git a/tools/build/src/engine/command.h b/tools/build/src/engine/command.h
index 9091cdcc74..0b968de589 100644
--- a/tools/build/src/engine/command.h
+++ b/tools/build/src/engine/command.h
@@ -43,7 +43,7 @@
 #include "config.h"
 #include "lists.h"
 #include "rules.h"
-#include "strings.h"
+#include "jam_strings.h"


 typedef struct _cmd CMD;
diff --git a/tools/build/src/engine/compile.cpp b/tools/build/src/engine/compile.cpp
index 6ebb4006c6..030d2e1865 100644
--- a/tools/build/src/engine/compile.cpp
+++ b/tools/build/src/engine/compile.cpp
@@ -33,7 +33,7 @@
 #include "parse.h"
 #include "rules.h"
 #include "search.h"
-#include "strings.h"
+#include "jam_strings.h"
 #include "variable.h"
 #include "output.h"

diff --git a/tools/build/src/engine/debugger.cpp b/tools/build/src/engine/debugger.cpp
index 06f9db5181..66aa85cddc 100644
--- a/tools/build/src/engine/debugger.cpp
+++ b/tools/build/src/engine/debugger.cpp
@@ -7,7 +7,7 @@

 #include "debugger.h"
 #include "constants.h"
-#include "strings.h"
+#include "jam_strings.h"
 #include "pathsys.h"
 #include "cwd.h"
 #include "function.h"
diff --git a/tools/build/src/engine/execcmd.h b/tools/build/src/engine/execcmd.h
index 59cba42457..bea3df6e3d 100644
--- a/tools/build/src/engine/execcmd.h
+++ b/tools/build/src/engine/execcmd.h
@@ -17,7 +17,7 @@

 #include "config.h"
 #include "lists.h"
-#include "strings.h"
+#include "jam_strings.h"
 #include "timestamp.h"


diff --git a/tools/build/src/engine/execunix.cpp b/tools/build/src/engine/execunix.cpp
index 98aad81343..26076dff90 100644
--- a/tools/build/src/engine/execunix.cpp
+++ b/tools/build/src/engine/execunix.cpp
@@ -10,7 +10,7 @@

 #include "lists.h"
 #include "output.h"
-#include "strings.h"
+#include "jam_strings.h"

 #include <errno.h>
 #include <signal.h>
diff --git a/tools/build/src/engine/filent.cpp b/tools/build/src/engine/filent.cpp
index 12b3c81fc2..717508ba0d 100644
--- a/tools/build/src/engine/filent.cpp
+++ b/tools/build/src/engine/filent.cpp
@@ -34,7 +34,7 @@

 #include "object.h"
 #include "pathsys.h"
-#include "strings.h"
+#include "jam_strings.h"
 #include "output.h"

 #ifdef __BORLANDC__
diff --git a/tools/build/src/engine/filesys.cpp b/tools/build/src/engine/filesys.cpp
index b92f1ef82d..f64eeb28e5 100644
--- a/tools/build/src/engine/filesys.cpp
+++ b/tools/build/src/engine/filesys.cpp
@@ -33,7 +33,7 @@
 #include "lists.h"
 #include "object.h"
 #include "pathsys.h"
-#include "strings.h"
+#include "jam_strings.h"
 #include "output.h"

 #include <assert.h>
diff --git a/tools/build/src/engine/fileunix.cpp b/tools/build/src/engine/fileunix.cpp
index 5f6815725f..50c2c67d67 100644
--- a/tools/build/src/engine/fileunix.cpp
+++ b/tools/build/src/engine/fileunix.cpp
@@ -33,7 +33,7 @@

 #include "object.h"
 #include "pathsys.h"
-#include "strings.h"
+#include "jam_strings.h"
 #include "output.h"

 #include <assert.h>
diff --git a/tools/build/src/engine/filevms.cpp b/tools/build/src/engine/filevms.cpp
index ef93b0cdb4..ab038d065b 100644
--- a/tools/build/src/engine/filevms.cpp
+++ b/tools/build/src/engine/filevms.cpp
@@ -18,7 +18,7 @@

 #include "object.h"
 #include "pathsys.h"
-#include "strings.h"
+#include "jam_strings.h"


 #ifdef OS_VMS
diff --git a/tools/build/src/engine/function.h b/tools/build/src/engine/function.h
index fb951c096e..73c837f520 100644
--- a/tools/build/src/engine/function.h
+++ b/tools/build/src/engine/function.h
@@ -12,7 +12,7 @@
 #include "frames.h"
 #include "lists.h"
 #include "parse.h"
-#include "strings.h"
+#include "jam_strings.h"

 typedef struct _function FUNCTION;
 typedef struct _stack STACK;
diff --git a/tools/build/src/engine/hdrmacro.cpp b/tools/build/src/engine/hdrmacro.cpp
index a8be7787dc..9b6507ec6a 100644
--- a/tools/build/src/engine/hdrmacro.cpp
+++ b/tools/build/src/engine/hdrmacro.cpp
@@ -36,7 +36,7 @@
 #include "object.h"
 #include "parse.h"
 #include "rules.h"
-#include "strings.h"
+#include "jam_strings.h"
 #include "subst.h"
 #include "variable.h"
 #include "output.h"
diff --git a/tools/build/src/engine/jam.cpp b/tools/build/src/engine/jam.cpp
index 850b30e01f..ef69e22f19 100644
--- a/tools/build/src/engine/jam.cpp
+++ b/tools/build/src/engine/jam.cpp
@@ -130,7 +130,7 @@
 #include "rules.h"
 #include "scan.h"
 #include "search.h"
-#include "strings.h"
+#include "jam_strings.h"
 #include "timestamp.h"
 #include "variable.h"
 #include "execcmd.h"
diff --git a/tools/build/src/engine/strings.cpp b/tools/build/src/engine/jam_strings.cpp
similarity index 99%
rename from tools/build/src/engine/strings.cpp
rename to tools/build/src/engine/jam_strings.cpp
index 9c100d87f0..01d39469d3 100644
--- a/tools/build/src/engine/strings.cpp
+++ b/tools/build/src/engine/jam_strings.cpp
@@ -3,7 +3,7 @@
 /* file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt) */

 #include "jam.h"
-#include "strings.h"
+#include "jam_strings.h"

 #include <assert.h>
 #include <stdlib.h>
diff --git a/tools/build/src/engine/strings.h b/tools/build/src/engine/jam_strings.h
similarity index 92%
rename from tools/build/src/engine/strings.h
rename to tools/build/src/engine/jam_strings.h
index 86cd1904b9..f47db10af7 100644
--- a/tools/build/src/engine/strings.h
+++ b/tools/build/src/engine/jam_strings.h
@@ -5,8 +5,8 @@
  * http://www.boost.org/LICENSE_1_0.txt)
  */

-#ifndef STRINGS_DWA20011024_H
-#define STRINGS_DWA20011024_H
+#ifndef JAM_STRINGS_DWA20011024_H
+#define JAM_STRINGS_DWA20011024_H

 #include "config.h"
 #include <stddef.h>
diff --git a/tools/build/src/engine/modules.cpp b/tools/build/src/engine/modules.cpp
index ca3e18ee2c..d30c40bc4d 100644
--- a/tools/build/src/engine/modules.cpp
+++ b/tools/build/src/engine/modules.cpp
@@ -14,7 +14,7 @@
 #include "object.h"
 #include "parse.h"
 #include "rules.h"
-#include "strings.h"
+#include "jam_strings.h"
 #include "variable.h"

 #include <assert.h>
diff --git a/tools/build/src/engine/modules/order.cpp b/tools/build/src/engine/modules/order.cpp
index 467322da0e..c5ebeba06a 100644
--- a/tools/build/src/engine/modules/order.cpp
+++ b/tools/build/src/engine/modules/order.cpp
@@ -8,7 +8,7 @@
 #include "../mem.h"
 #include "../native.h"
 #include "../object.h"
-#include "../strings.h"
+#include "../jam_strings.h"
 #include "../variable.h"


diff --git a/tools/build/src/engine/modules/regex.cpp b/tools/build/src/engine/modules/regex.cpp
index e6a7bc9a7c..fa53558588 100644
--- a/tools/build/src/engine/modules/regex.cpp
+++ b/tools/build/src/engine/modules/regex.cpp
@@ -7,7 +7,7 @@

 #include "../mem.h"
 #include "../native.h"
-#include "../strings.h"
+#include "../jam_strings.h"
 #include "../subst.h"

 /*
diff --git a/tools/build/src/engine/pathsys.h b/tools/build/src/engine/pathsys.h
index 6c84396d1f..839476e940 100644
--- a/tools/build/src/engine/pathsys.h
+++ b/tools/build/src/engine/pathsys.h
@@ -23,7 +23,7 @@

 #include "config.h"
 #include "object.h"
-#include "strings.h"
+#include "jam_strings.h"


 typedef struct _pathpart
diff --git a/tools/build/src/engine/search.cpp b/tools/build/src/engine/search.cpp
index e21ae3687e..7529e02dcb 100644
--- a/tools/build/src/engine/search.cpp
+++ b/tools/build/src/engine/search.cpp
@@ -20,7 +20,7 @@
 #include "lists.h"
 #include "object.h"
 #include "pathsys.h"
-#include "strings.h"
+#include "jam_strings.h"
 #include "timestamp.h"
 #include "variable.h"
 #include "output.h"
diff --git a/tools/build/src/engine/timestamp.cpp b/tools/build/src/engine/timestamp.cpp
index 2add949f97..c998730bed 100644
--- a/tools/build/src/engine/timestamp.cpp
+++ b/tools/build/src/engine/timestamp.cpp
@@ -31,7 +31,7 @@
 #include "hash.h"
 #include "object.h"
 #include "pathsys.h"
-#include "strings.h"
+#include "jam_strings.h"
 #include "output.h"


diff --git a/tools/build/src/engine/variable.cpp b/tools/build/src/engine/variable.cpp
index cdb8a32da3..574c344ce6 100644
--- a/tools/build/src/engine/variable.cpp
+++ b/tools/build/src/engine/variable.cpp
@@ -38,7 +38,7 @@
 #include "modules.h"
 #include "parse.h"
 #include "pathsys.h"
-#include "strings.h"
+#include "jam_strings.h"
 #include "output.h"

 #include <stdio.h>
diff --git a/tools/build/src/engine/w32_getreg.cpp b/tools/build/src/engine/w32_getreg.cpp
index dd2d0fc70c..15c0c947d0 100644
--- a/tools/build/src/engine/w32_getreg.cpp
+++ b/tools/build/src/engine/w32_getreg.cpp
@@ -12,7 +12,7 @@ file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
 # include "object.h"
 # include "parse.h"
 # include "frames.h"
-# include "strings.h"
+# include "jam_strings.h"

 # define WIN32_LEAN_AND_MEAN
 # include <windows.h>
